lms/djangoapps/grades/models.py

```diff
@@ -311,74 +262,98 @@ def read_grade(cls, user_id, usage_key):
-            course_version,
-            subtree_edited_timestamp,
-            earned_all,
-            possible_all,
-            earned_graded,
-            possible_graded,
-            visible_blocks,
-    ):
-        """
-        Updates a previously existing grade.
-
-        This is distinct from update() in that `grade.update()` operates on an
-        existing grade object, while this is a classmethod that pulls the grade
-        from the database, and then updates it.  If you already have a grade
-        object, use the update() method on that object to avoid an extra
-        round-trip to the database.  Use this classmethod if all you have are a
-        user and the usage key of an existing grade.
-
-        Requires all the arguments listed in docstring for create_grade
-        """
-        grade = cls.read_grade(
+    def bulk_read_grades(cls, user_id, course_key):
+        """
+        Reads all grades for the given user and course.
+
+        Arguments:
+            user_id: The user associated with the desired grades
+            course_key: The course identifier for the desired grades
+        """
+        return cls.objects.select_related('visible_blocks').filter(
+            user_id=user_id,
+            course_id=course_key,
+        )
+
+    @classmethod
+    def update_or_create_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of those methods.
+        """
+        cls._prepare_params_and_visible_blocks(kwargs)
+
+        user_id = kwargs.pop('user_id')
+        usage_key = kwargs.pop('usage_key')
+
+        grade, _ = cls.objects.update_or_create(
             user_id=user_id,
+            course_id=usage_key.course_key,
             usage_key=usage_key,
+            defaults=kwargs,
         )
+        return grade
 
-        grade.update(
-            course_version=course_version,
-            subtree_edited_timestamp=subtree_edited_timestamp,
-            earned_all=earned_all,
-            possible_all=possible_all,
-            earned_graded=earned_graded,
-            possible_graded=possible_graded,
-            visible_blocks=visible_blocks,
+    @classmethod
+    def create_grade(cls, **kwargs):
+        """
+        Wrapper for objects.create.
+        """
+        cls._prepare_params_and_visible_blocks(kwargs)
+        return cls.objects.create(**kwargs)
+
+    @classmethod
+    def bulk_create_grades(cls, grade_params_iter, course_key):
+        """
+        Bulk creation of grades.
+        """
+        if not grade_params_iter:
+            return
+
+        map(cls._prepare_params, grade_params_iter)
+        VisibleBlocks.bulk_get_or_create([params['visible_blocks'] for params in grade_params_iter], course_key)
+        map(cls._prepare_params_visible_blocks_id, grade_params_iter)
+
+        return cls.objects.bulk_create([PersistentSubsectionGrade(**params) for params in grade_params_iter])
```

Comment:
As discussed offline, we'll be taking the approach for a catch-all race-condition handling as explained here: https://openedx.atlassian.net/browse/TNL-5503